### PR TITLE
feat: dnsping using step-by-step

### PR DESCRIPTION
**Pull request by DecFox**
## Checklist

- [x] I have read the [contribution guidelines](https://github.com/ooni/probe-cli/blob/master/CONTRIBUTING.md)
- [x] reference issue for this pull request: [Issue #2159](https://github.com/ooni/probe/issues/2159)
- [x] if you changed anything related how experiments work and you need to reflect these changes in the ooni/spec repository, please link to the related ooni/spec pull request: https://github.com/ooni/spec/pull/252

## Description

This diff refactors the `dnsping` experiment to use the step-by-step measurement style ([doc here](https://github.com/ooni/probe-cli/blob/master/docs/design/dd-003-step-by-step.md))

### CODE REVIEW

File: internal/model/mocks/trace.go [hunk 1]
```diff
@@ -15,6 +15,9 @@ import (
 type Trace struct {
 	MockTimeNow func() time.Time
 
+	MockOnDNSRoundTripForLookupHost func(
+		started time.Time, reso model.Resolver, query model.DNSQuery, response model.DNSResponse,
+		addrs []string, err error, finished time.Time)
```
Comment (bassosimone) [inline internal/model/mocks/trace.go:20]:
```suggestion
		addrs []string, err error, finished time.Time)

```

For aesthetic reasons I'd like to have an empty line here

File: internal/model/netx.go [hunk 2]
```diff
@@ -303,6 +303,26 @@ type Trace interface {
 	// can use functionality exported by the ./internal/testingx pkg.
 	TimeNow() time.Time
 
+	// OnDNSLookupDone is called when DNSLookup terminates
```
Comment (bassosimone) [inline internal/model/netx.go:306]: Beware that the function name has changed below. I think you should also clarify that this gets called when you are using a transport as opposed to another function that we'll need for getaddrinfo.

I think the very important take home message here is that this callback is called after a single round trip.

File: internal/model/netx.go [hunk 3]
```diff
@@ -303,6 +303,26 @@ type Trace interface {
...
+	//
+	// Arguments:
```
Comment (bassosimone) [inline internal/model/netx.go:308]: Could you please document which arguments could be nil? (If other functions above do not do that, please apply the boy scout rule and ensure all other functions in this file document which arguments could be nil)

File: internal/engine/experiment/dnsping/dnsping.go [hunk 4]
```diff
@@ -119,11 +120,10 @@ func (m *Measurer) Run(
 	}
 	tk := new(TestKeys)
 	measurement.TestKeys = tk
-	mxmx := measurex.NewMeasurerWithDefaultSettings()
-	out := make(chan *measurex.DNSMeasurement)
+	out := make(chan *SinglePing)
```
Comment (bassosimone) [inline internal/engine/experiment/dnsping/dnsping.go:123]: Above there's a TODO comment for me that now has been done by your PR, so please zap this comment

File: internal/engine/experiment/dnsping/dnsping.go [hunk 5]
```diff
@@ -132,54 +132,54 @@ func (m *Measurer) Run(
...
-// onlyQueryWithType returns only the queries with the given type.
-func (m *Measurer) onlyQueryWithType(
-	in []*measurex.ArchivalDNSLookupEvent, kind string) (out []*SinglePing) {
-	for _, query := range in {
-		if query.QueryType == kind {
-			out = append(out, &SinglePing{
-				Queries: []*measurex.ArchivalDNSLookupEvent{query},
+		for _, query := range queries {
+			tk.Pings = append(tk.Pings, &SinglePing{
+				Queries: []*model.ArchivalDNSLookupResult{query},
```
Comment (bassosimone) [inline internal/engine/experiment/dnsping/dnsping.go:142]: Would it be possible to fill `tk.Pings` at line `134`? I think we don't need to split the queries anymore, after your changes, so also my TODO comment can now be removed. I think this whole loop at the end is now unnecessary.

**Review by bassosimone — COMMENTED** [review]
Preliminary review: this is a very good start, keep going! 💯 

I have left comments regarding the structure of the code and how to improve the documentation. It seems the structure is good and there's just some space for applying small improvements along the way. I didn't look into tests at this stage.

File: internal/measurexlite/trace.go [hunk 6]
```diff
@@ -38,6 +39,10 @@ type Trace struct {
 	// this channel manually, ensure it has some buffer.
 	NetworkEvent chan *model.ArchivalNetworkEvent
 
+	// NewParallelResolverFn is OPTIONAL and can be used to overide
```
[the remark this answers is not in the document]
Comment (DecFox) [reply internal/measurexlite/trace.go:42]: Thanks for pointing this out!

File: internal/measurexlite/trace.go [hunk 7]
```diff
@@ -110,14 +127,23 @@ func NewTrace(index int64, zeroTime time.Time) *Trace {
 }
 
 // newDialerWithoutResolver indirectly calls netxlite.NewDialerWithoutResolver
-// thus allows us to mock this func for testing.
+// thus allowing us to mock this func for testing.
```
[the remark this answers is not in the document]
Comment (DecFox) [reply internal/measurexlite/trace.go:133]: I am not entirely sure (I am probably not the best judge here either). The function below used "allowing", so I settled for it (randomly)

**Comment by DecFox** [comment]
Thanks for the feedback! Made the required changes.

**Comment by DecFox** [comment]
We have a test failure here:
```
--- FAIL: TestNewTLSHandshakerStdlib (1.27s)
    --- FAIL: TestNewTLSHandshakerStdlib/Handshake_saves_into_the_trace (0.00s)
        tls_test.go:108: unexpected err unknown_failure: mocked
FAIL
coverage: 90.9% of statements
```
Here: https://github.com/ooni/probe-cli/runs/7236058975?check_suite_focus=true#step:4:168

**Comment by DecFox** [comment]
We had a similar test failure signifying that the fix was not good:
```
 --- FAIL: TestNewTLSHandshakerStdlib (2.23s)
    --- FAIL: TestNewTLSHandshakerStdlib/Handshake_discards_events_when_buffers_are_full (0.00s)
        tls_test.go:219: unexpected err unknown_failure: mocked
FAIL
coverage: 90.9% of statements
```
This is probably because we have a cancelled context and `<-ctx.Done` returns immediately

The failure was subsequently fixed by @bassosimone (thanks!) in https://github.com/ooni/probe-cli/pull/839.

File: internal/engine/experiment/dnsping/dnsping.go (hunk 5 above)
Comment (DecFox) [reply internal/engine/experiment/dnsping/dnsping.go]: Introducing the new experiment pattern which helps with this: https://github.com/ooni/probe-cli/pull/831/commits/2ee2eb906e91604b39c1c9985953af8d56b05725

File: internal/model/netx.go (hunk 3 above)
Comment (DecFox) [reply internal/model/netx.go:309]: A little improvement on the docs. Maybe this grasps what we expect in arguments better: 
https://github.com/ooni/probe-cli/pull/831/commits/05369a38b3a7a3d71e1f71c1b09a773ad477a228#diff-f9f42daa515ddf8adbc780a80668686e9e2425baff569b86c38e01c1f2449636

Comment (bassosimone) [reply internal/model/netx.go:309]: Yeah!

File: internal/engine/experiment/dnsping/dnsping.go [hunk 8]
```diff
@@ -117,69 +105,58 @@ func (m *Measurer) Run(
...
 // dnsRoundTrip performs a round trip and returns the results to the caller.
-func (m *Measurer) dnsRoundTrip(ctx context.Context, mxmx *measurex.Measurer,
-	address string, domain string) *measurex.DNSMeasurement {
+func (m *Measurer) dnsRoundTrip(ctx context.Context, index int64, zeroTime time.Time,
+	logger model.Logger, address string, domain string, wg *sync.WaitGroup, tk *TestKeys) {
 	// TODO(bassosimone): make the timeout user-configurable
 	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
 	defer cancel()
-	return mxmx.LookupHostUDP(ctx, domain, address)
+
```
Comment (bassosimone) [inline internal/engine/experiment/dnsping/dnsping.go:139]:
```suggestion
```

File: internal/measurexlite/dns_test.go [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+			}
+		})
+	})
+
+	t.Run("LookupHost discards events when buffers are full", func(t *testing.T) {
+		zeroTime := time.Now()
+		td := testingx.NewTimeDeterministic(zeroTime)
+		trace := NewTrace(0, zeroTime)
+		trace.DNSLookup = map[uint16]chan *model.ArchivalDNSLookupResult{
+			dns.TypeA:    make(chan *model.ArchivalDNSLookupResult),
```
Comment (bassosimone) [inline internal/measurexlite/dns_test.go:170]:
```suggestion
			dns.TypeA:    make(chan *model.ArchivalDNSLookupResult), // no buffer
```

File: internal/measurexlite/dns_test.go [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+			dns.TypeAAAA: make(chan *model.ArchivalDNSLookupResult),
```
Comment (bassosimone) [inline internal/measurexlite/dns_test.go:171]:
```suggestion
			dns.TypeAAAA: make(chan *model.ArchivalDNSLookupResult), // no buffer
```

File: internal/measurexlite/dns_test.go [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+	}, {
+		name: "with invalid IPv4 address",
+		args: []string{"1.1.1.1.1", "fe80::a00:20ff:feb9:4c54"},
+	}, {
+		name: "with invalid IPv6 address",
+		args: []string{"1.1.1.1", "fe80::a00:20ff:feb9:::4c54"},
+	}, {
+		name: "with empty input",
+		args: []string{},
+	}}
```
Comment (bassosimone) [inline internal/measurexlite/dns_test.go:240]: @DecFox please add here a case `"with nil input"` where `args: nil`

Comment (bassosimone) [inline internal/measurexlite/dns_test.go:240]:
```suggestion
	}, {
		name: "with nil input",
		args: nil,
	}
```

File: internal/measurexlite/trace.go [hunk 12]
```diff
@@ -46,6 +51,11 @@ type Trace struct {
 	// calls to the netxlite.NewTLSHandshakerStdlib factory.
 	NewTLSHandshakerStdlibFn func(dl model.DebugLogger) model.TLSHandshaker
 
+	// DNSLookup is MANDATORY and buffers DNSLookup results based on the query type
+	// If you create this channel manually, ensure all channels corresponding
+	// to the queries of interest have some buffer
```
Comment (bassosimone) [inline internal/measurexlite/trace.go:56]:
```suggestion
	// to the queries of interest have some buffer.
```

File: internal/measurexlite/trace.go [hunk 13]
```diff
@@ -46,6 +51,11 @@ type Trace struct {
 	// calls to the netxlite.NewTLSHandshakerStdlib factory.
 	NewTLSHandshakerStdlibFn func(dl model.DebugLogger) model.TLSHandshaker
 
+	// DNSLookup is MANDATORY and buffers DNSLookup results based on the query type
```
Comment (bassosimone) [inline internal/measurexlite/trace.go:54]:
```suggestion
	// DNSLookup is MANDATORY and buffers DNSLookup results based on the query type.
```

File: internal/measurexlite/trace.go [hunk 14]
```diff
@@ -96,6 +110,12 @@ func NewTrace(index int64, zeroTime time.Time) *Trace {
 		),
 		NewDialerWithoutResolverFn: nil, // use default
 		NewTLSHandshakerStdlibFn:   nil, // use default
+		DNSLookup: map[uint16]chan *model.ArchivalDNSLookupResult{
+			dns.TypeANY:   make(chan *model.ArchivalDNSLookupResult, DNSLookupBufferSize),
+			dns.TypeA:     make(chan *model.ArchivalDNSLookupResult, DNSLookupBufferSize),
+			dns.TypeAAAA:  make(chan *model.ArchivalDNSLookupResult, DNSLookupBufferSize),
+			dns.TypeCNAME: make(chan *model.ArchivalDNSLookupResult, DNSLookupBufferSize),
```
Comment (bassosimone) [inline internal/measurexlite/trace.go:117]:
```suggestion
			dns.TypeCNAME: make(chan *model.ArchivalDNSLookupResult, DNSLookupBufferSize),
			dns.TypeNS: make(chan *model.ArchivalDNSLookupResult, DNSLookupBufferSize),
```

File: internal/measurexlite/trace_test.go [hunk 15]
```diff
@@ -111,6 +153,57 @@ func TestNewTrace(t *testing.T) {
 }
 
 func TestTrace(t *testing.T) {
+	t.Run("NewUnwrappedParallelResolverFn works as intended", func(t *testing.T) {
```
Comment (bassosimone) [inline internal/measurexlite/trace_test.go:156]:
```suggestion
	t.Run("NewParallelResolverFn works as intended", func(t *testing.T) {
```

File: internal/model/netx.go [hunk 16]
```diff
@@ -303,6 +303,28 @@ type Trace interface {
...
+	// - started is when we called transport.RoundTrip
+	//
+	// - reso is the parent resolver for the trace;
+	//
+	// - query is the non-nil DNS query we use for the RoundTrip
+	//
+	// - response is a valid DNS response, obtained after the RoundTrip;
+	//
+	// - addrs is the list of addresses obtained after the RoundTrip, which
+	// is empty if the RoundTrip faled
```
Comment (bassosimone) [inline internal/model/netx.go:320]:
```suggestion
	// is empty if the RoundTrip failed
```

File: internal/netxlite/trace.go [hunk 17]
```diff
@@ -49,6 +49,11 @@ func (*traceDefault) TimeNow() time.Time {
 	return time.Now()
 }
 
+func (*traceDefault) OnDNSRoundTripForLookupHost(started time.Time, reso model.Resolver, query model.DNSQuery,
```
Comment (bassosimone) [inline internal/netxlite/trace.go:52]:
```suggestion
// OnDNSRoundTripForLookupHost implements model.Trace.OnDNSRoundTripForLookupHost.
func (*traceDefault) OnDNSRoundTripForLookupHost(started time.Time, reso model.Resolver, query model.DNSQuery,
```

**Review by bassosimone — CHANGES_REQUESTED** [review]
You did a very good job here, @DecFox! 💯 🔝 

I applied [some changes](https://github.com/ooni/probe-cli/compare/acc3d0c1742d399d6bdc88ee6e0ea76456f952b5...f56de72a2f3a4c2e730ffefe21593088d3744080), which I kindly ask you to double check. It seems, in particular, I did remember incorrectly the [Go channel axioms](https://dave.cheney.net/2014/03/19/channel-axioms): we were not risking a panic rather blocking forever. I suggest you also take a look at Go channel axioms, if you're not already familiar with them.

I was hoping of merging after I applied these changes, but then I realized we still actually need to apply the following changes:

1. please update https://github.com/ooni/spec because we have just changed the data format and we need to reflect that inside the experiment's spec;

2. please bump the experiment's version number's minor version and reset the patch version because we have just changed the data format, so we need to do that (I think there is also a check in the tests that needs to be updated).

After all of this happens, for me it will become good to merge into master.

**Review by bassosimone — APPROVED** [review]
🐳 🙏 🥳

**Pull request merged**